### ISSUE TITLE
API - PHP How to avoid false positive ?

### ISSUE THREAD

**Issue by cyvax** [PrivateBin/PrivateBin#716]
I'm working on the api using php, I have a code that works but at times I have "false positives" that doesn't work ... 

someone would know how to fix it ? 

(I call "false positives", the paste that asks me for a password)
<details>
  <summary>code</summary>

```php
<?php
use Tuupola\Base58;

$base58 = new Base58(["characters" => Base58::BITCOIN]);


$iv = random_bytes(16);
$salt = random_bytes(8);
$password = random_bytes(32);
$b58 = $base58->encode($password);
$aad = [
    [
        base64_encode($iv),
        base64_encode($salt),
        100000,
        256,
        128,
        "aes",
        "gcm",
        "none"
    ],
    "markdown",
    0,
    0];

$key = openssl_pbkdf2($password, $salt, 32, 100000, 'sha256');
$paste = ['paste' => 'text content of the paste'];
$crypt = openssl_encrypt(json_encode($paste), 'aes-256-gcm', $key, $options=OPENSSL_RAW_DATA, $iv, $tag, json_encode($aad), 16.0); 

$data = array(
            "v" => 2,
            "adata" => $aad,
            "ct" => base64_encode($crypt . $tag),
            "meta" => array(
                "expire" => "5min"
            )
    );

$url = "https://privatebin.net/";
$ch = curl_init($url);
curl_setopt($ch, CURLOPT_POST, 1);
curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));
curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
curl_setopt($ch, CURLOPT_HTTPHEADER, array(
    'Content-Type: application/json',
    'X-Requested-With: JSONHttpRequest'));
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true); 
$result = json_decode(curl_exec($ch));
curl_close($ch);
echo $url . $result->url . "#" . $b58;
?>
```
</details>

**Comment by rugk** [comment]
If you get a password request if you actually did not set a password, this means the decryption fails. Specifically see https://github.com/PrivateBin/PrivateBin/wiki/Encryption-format for the encryption format you need to use and [examples here for other third-party clients](https://github.com/PrivateBin/PrivateBin/wiki/Third-party-clients).

In your case, you seem to encrypt it somehow via OpenSSL and I guess the result is just not what we expect.
If this question has been answered and feel free to close this issue. If anyone has further questions on that topic, feel free to comment here, again.

**Comment by cyvax** [comment]
I understand, but it actually works, but sometimes the password prompt appears but not always, that's what I don't understand...

**Comment by rugk** [comment]
Yeah, don't know why that would happen, too…

**Comment by elrido** [comment]
I would recommend to set up some loop to generate 100 or so at a time and then start to fix the keys and other variables. Basically only generate one of the inputs randomly at a time (i.e. fixed key, iv, salt, but random paste), so that you can figure out where the issue is.

We had a for example an issue with when we switched to base58, which trims leading zero bytes. That caused the decryption to fail when the random key happened to start with one or multiple 0x00 bytes. It could be a similar effect in one of these functions that trims "empty" values at the beginning or end of your primitives.

Divide and conquer to isolate which one it is.

**Comment by cyvax** [comment]
I found the error...

after following the tracks left by elrido, i used the api to retrieve the paste (wanted to use python to check if the paste was valid or not), and i noticed a similarity...
here some iv and salt of false positive... : 
| iv  | salt |
| ------------- | ------------- |
| W5jGqYHOEp+fxYZjHNF\\/TA== | HxwySup3b1M= |
| W+9ykprin\\/JIL3x7sGwp9Q== | xOzgMa+8+Qg= |
| z2DJmiO+4FLzNUWvCe+nJg== | YikyOL\\/kcow= |
| vxIe6qE7BwhaN7Jn5D\\/rwQ== | dOVtyD3AgbM= |
| yWU8vSlhUohC\\/AiUUZcA8w== | nqnKBgEuPTQ= |
| WRai9ScJZXretdnLIWB4bA== | MqqZmjNt\\/MI= |

yes the error was caused by json_encode...
by adding : JSON_UNESCAPED_SLASHES on the json_encode it works now...

json_encode($data, JSON_UNESCAPED_SLASHES)

thanks anyway i learn something new today ^^
<details>
  <summary>working code</summary>

```php
<?php
use Tuupola\Base58;

$base58 = new Base58(["characters" => Base58::BITCOIN]);


$iv = random_bytes(16);
$salt = random_bytes(8);
$password = random_bytes(32);
$b58 = $base58->encode($password);
$aad = [
    [
        base64_encode($iv),
        base64_encode($salt),
        100000,
        256,
        128,
        "aes",
        "gcm",
        "none"
    ],
    "markdown",
    0,
    0];

$key = openssl_pbkdf2($password, $salt, 32, 100000, 'sha256');
$paste = ['paste' => 'text content of the paste'];
$crypt = openssl_encrypt(json_encode($paste), 'aes-256-gcm', $key, $options=OPENSSL_RAW_DATA, $iv, $tag, json_encode($aad, JSON_UNESCAPED_SLASHES), 16.0);

$data = array(
            "v" => 2,
            "adata" => $aad,
            "ct" => base64_encode($crypt . $tag),
            "meta" => array(
                "expire" => "5min"
            )
    );

$url = "https://privatebin.net/";
$ch = curl_init($url);
curl_setopt($ch, CURLOPT_POST, 1);
curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data, JSON_UNESCAPED_SLASHES));
curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
curl_setopt($ch, CURLOPT_HTTPHEADER, array(
    'Content-Type: application/json',
    'X-Requested-With: JSONHttpRequest'));
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true); 
$result = json_decode(curl_exec($ch));
curl_close($ch);
echo $url . $result->url . "#" . $b58;
?>
```
</details>

**Comment by rugk** [comment]
BTW if you want to create/created an PHP client for PrivateBin, feel free to libnk it/add it to the wiki at https://github.com/PrivateBin/PrivateBin/wiki/Third-party-clients. Thanks. :smiley:

**Comment by cyvax** [comment]
Of course I will add it now  ^^

**Comment by r4sas** [comment]
I'm glad you managed to implement sending pastes using PHP.
Earlier I tried to do that here: https://github.com/PrivateBin/PrivateBin/issues/526#issuecomment-536994871

**Comment by cyvax** [comment]
Hey!

I've seen your tests, I had already tested a while ago to correct it but I was running out of time so I went back to NodeJS, but I'm redoing my backend so I'm looking at it again.

I have already implemented some of this in my library.
I was ~~slightly~~ Heavily inspired by PBinCLI (especially on how to manage data properly like zlib, paste password etc).

so well we have a Privatebin_PHP library now (with terrible documentation because i suck at writing them). 
^^
